### ISSUE TITLE
How to dynamically change the network layer? 

### ISSUE THREAD

**Issue by ron-liu** [facebook/relay#1215]
How to call injectNetworkLayer without token when bootstrapping, and then overwrite the networklayer after signing in, like the following code shows:

```javascript
// when bootstrapping
Relay.injectNetworkLayer(
    new Relay.DefaultNetworkLayer('http://some-server/graphql')
);

// after signing in
signIn().then(token => {
   Relay.injectNetworkLayer(
    new Relay.DefaultNetworkLayer('/graphql'), {
        headers: {
            Authorization: `Bearer  ${token}`
        }
    } 
   // Currently, the above sentence will throw exception and prevent overwriting, what is the other way to accomplish?  

});
```

**Comment by nodkz** [comment]
Very strange, but it should work.
Version 0.8.0 send noisy warning to console, when Default network layer was reinjected. But all works properly. Try to update till 0.8.1 or 0.9.0.

If it does not help, you may try my OSS implementation of network layer https://github.com/nodkz/react-relay-network-layer It contains several good middlewares, which solves common tasks.

So your code may be implemented such way:
```js

const tokenInGlobalVars = null;

signIn().then(token => {
   tokenInGlobalVars = token;
});

Relay.injectNetworkLayer(new RelayNetworkLayer([
  urlMiddleware({
    url: (req) => 'http://some-server/graphql', // this method will be called every time, so you may change url on fly
  }),
  authMiddleware({
    token: () => tokenInGlobalVars,
    allowEmptyToken: true,
  }),
], { disableBatchQuery: true }));
```

Also was discussion where and how to store a token on the client side: https://github.com/nodkz/react-relay-network-layer/issues/6

**Comment by ron-liu** [comment]
I already in 0.9.0, I am sure when calling injectNetworkLayer second time, it will cause an exception.

@nodkz, thanks for your powerful library, I will use it soon.

**Comment by wincent** [comment]
I'd recommend using @nodkz's layer like he describes. The intention is to not inject new network layers all the time, but rather inject a single layer that is flexible enough to embody your desired behavior. (We made this stricter in response to bugs that cropped up internally in large apps where we were trying to update the layer in multiple places and unknowingly clobbering previous adjustments.)

As I think this is now resolved, going to mark it as closed.
